include/deal.II/fe/fe_enriched.h

```diff
@@ -0,0 +1,519 @@
+ *
+ * <h3>hp-constraints</h3>
+ *
+ * This class implements hp constraints with arbitrary FE. For the example above,
+ * a user will need to wrap a given FiniteElement into FE_Enriched:
+ * @code
+ *   FE_Enriched<dim> fe_enriched(FE_Q<dim>(1));
+ * @endcode
+ *
+ * This corresponds to building constraints between enriched FE and
+ *
+ * @code
+ * FESystem<dim>(FE_Q<dim>(1),1,
+ *               FE_Nothing<dim>(1,true),1)
+ * @endcode
+ *
+ * In order to make the resulting field continuous
+ * all enriched DoFs on the interface are constrained to be zero.
+ * This is a natural outcome of FE_Nothing with domination set to <code>true</code>.
+ *
+ * Note, that if the constraints are built between two different instances of this class,
+ * no checks are performed to ensure that the underlying enrichment function is
+ * the same. Thus the user must be careful in order not to end up with the
+ * discontinuous FE space.
+ *
+ * When using this class, please cite
+ * @code{.bib}
+ @Article{Davydov2016,
+    author =    {Denis Davydov and Tymofiy Gerasimov and Jean-Paul Pelteret and Paul Steinmann},
+    title =     {On the h-adaptive PUM and hp-adaptive FEM approaches applied to PDEs in quantum mechanics.},
+    journal =   {Journal of Computational Physics. Submitted},
+    year =     {2016},
+ }
+ * @endcode
+ *
+ * @ingroup fe
+ *
+ * @author Denis Davydov, 2016.
+ */
+template <int dim, int spacedim=dim>
+class FE_Enriched : public FiniteElement<dim, spacedim>
+{
+public:
+
+  /**
+   * Constructor which takes base FiniteElement @p fe_base and the enrichment
+   * FiniteElement @p fe_enriched which will be multiplied by the @p enrichment_function.
+   */
+  FE_Enriched (const FiniteElement<dim,spacedim> &fe_base,
+               const FiniteElement<dim,spacedim> &fe_enriched,
+               const Function<dim>      *enrichment_function);
+
+  /**
+   * Constructor which only wraps the base FE @p fe_base.
+   * As for enriched FiniteElement, FE_Nothing is used.
+   * Continuity constraints will be automatically generated when
+   * this non-enriched element is used in conjunction with enriched finite element
+   * within the hp::DoFHandler.
+   */
+  FE_Enriched (const FiniteElement<dim,spacedim> &fe_base);
+
+
+  /**
+   * Constructor which takes pointer to the base FiniteElement @p fe_base and
+   * a vector of enriched FiniteElement's @p fe_enriched . @p i'th element of this
+   * vector will be enriched with @p i'th function in @p functions.
+   *
+   * @note When used with FE_Nothing finite element, the pointers to functions can be NULL
+   * as they will not be used.
+   */
+  FE_Enriched (const FiniteElement<dim,spacedim> *fe_base,
+               const std::vector<const FiniteElement<dim,spacedim> * > &fe_enriched,
+               const std::vector<std::vector<std::function<const Function<dim> *(const typename Triangulation<dim, spacedim>::cell_iterator &) > > > &functions);
```

Comment:
I assume the nested `vector<vector<...>` construction is that you want to add different functions for different cells? Maybe add a more verbose discussion in the description of the constructor. Also, I'm not completely clear on how the enrichment is combined when having multiple vectors...